File: packages/loom/lsp/src/hover.rs
//! Hover information for directive verbs, CHARACTER refs, and beat
//! diverts. Resolution is purely lexical: we identify the token under
//! the cursor on the source line and look it up in the workspace
//! index.

use lsp_types::{Hover, HoverContents, MarkupContent, MarkupKind, Position, Range, Url};

use crate::completion::DIRECTIVES;
use crate::workspace::{line_at, Workspace};

pub fn hover_at(ws: &Workspace, uri: &Url, pos: Position) -> Option<Hover> {
    let doc = ws.docs.get(uri)?;
    let line = line_at(&doc.text, pos.line);
    let (token, start, end) = token_at(line, pos.character as usize)?;
    let range = Range {
        start: Position {
            line: pos.line,
            character: start as u32,
        },
        end: Position {
            line: pos.line,
            character: end as u32,
        },
    };

    // Directive head: `<name…` token immediately follows `<`.
    if let Some(open) = line[..start].rfind('<') {
        if line[open + 1..start].trim().is_empty()
            && DIRECTIVES.contains(&token)
        {
            return Some(markdown(
                format!("`<{token}: …>` — directive"),
                Some(range),
            ));
        }
    }

    // Divert tail: `-> token` form.
    if let Some(arrow) = line[..start].rfind("->") {
        if line[arrow + 2..start].trim().is_empty() {
            if let Some(entries) = ws.beats.get(token) {
                if let Some(beat) = entries.first() {
                    let mut body = format!("**beat** `{token}`");
                    if let Some(cast) = &beat.cast {
                        body.push_str(&format!("\n\ncast: {cast}"));
                    }
                    if let Some(setting) = &beat.setting {
                        body.push_str(&format!("\n\nsetting: {setting}"));
                    }
                    return Some(markdown(body, Some(range)));
                }
            }
        }
    }

    // CHARACTER reference: ALL-CAPS token that matches a declared name.
    if token.chars().all(|c| c.is_ascii_uppercase() || c == '_') {
        if let Some(info) = ws.characters.get(token) {
            let mut body = format!("**CHARACTER** `{token}`");
            if !info.mixins.is_empty() {
                body.push_str(&format!("\n\nis {}", info.mixins.join(", ")));
            }
            if !info.body.is_empty() {
                body.push_str("\n\n```loom\n");
                for line in &info.body {
                    body.push_str(line);
                    body.push('\n');
                }
                body.push_str("```");
            }
            return Some(markdown(body, Some(range)));
        }
    }

    None
}

fn markdown(value: String, range: Option<Range>) -> Hover {
    Hover {
        contents: HoverContents::Markup(MarkupContent {
            kind: MarkupKind::Markdown,
            value,
        }),
        range,
    }
}

/// Identify the alphanumeric / `_` token under the byte offset `col`.
pub fn token_at(line: &str, col: usize) -> Option<(&str, usize, usize)> {
    if col > line.len() {
        return None;
    }
    let bytes = line.as_bytes();
    let is_word = |b: u8| b.is_ascii_alphanumeric() || b == b'_';
    let mut start = col;
    while start > 0 && is_word(bytes[start - 1]) {
        start -= 1;
    }
    let mut end = col;
    while end < bytes.len() && is_word(bytes[end]) {
        end += 1;
    }
    if start == end {
        return None;
    }
    Some((&line[start..end], start, end))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn token_in_middle() {
        let (tok, s, e) = token_at("    -> ringing", 10).unwrap();
        assert_eq!(tok, "ringing");
        assert_eq!((s, e), (7, 14));
    }
}

File: packages/loom/core/src/lsp/hover.ts
//! Hover information for directive verbs, CHARACTER refs, and beat
//! diverts. Resolution is purely lexical: we identify the token under
//! the cursor on the source line and look it up in the workspace index.

import { type Hover, type Position, type Range } from "./types.ts";
import { allChars, isAsciiUppercase } from "../parser/rust.ts";
import { DIRECTIVES } from "./completion.ts";
import { lineAt, tokenAt } from "./util.ts";
import type { Workspace } from "./workspace.ts";

export function hoverAt(ws: Workspace, uri: string, pos: Position): Hover | null {
  const doc = ws.docs.get(uri);
  if (!doc) return null;
  const line = lineAt(doc.text, pos.line);
  const found = tokenAt(line, pos.character);
  if (!found) return null;
  const [token, start, end] = found;
  const range: Range = {
    start: { line: pos.line, character: start },
    end: { line: pos.line, character: end },
  };

  // Directive head: `<name…` token immediately follows `<`.
  const open = line.slice(0, start).lastIndexOf("<");
  if (open >= 0) {
    if (line.slice(open + 1, start).trim() === "" && DIRECTIVES.includes(token)) {
      return markdown(`\`<${token}: …>\` — directive`, range);
    }
  }

  // Divert tail: `-> token` form.
  const arrow = line.slice(0, start).lastIndexOf("->");
  if (arrow >= 0) {
    if (line.slice(arrow + 2, start).trim() === "") {
      const entries = ws.beats.get(token);
      const beat = entries?.[0];
      if (beat) {
        let body = `**beat** \`${token}\``;
        if (beat.cast !== null) body += `\n\ncast: ${beat.cast}`;
        if (beat.setting !== null) body += `\n\nsetting: ${beat.setting}`;
        return markdown(body, range);
      }
    }
  }

  // TRAIT reference: a declared trait name (mixed-case, so it never collides
  // with the ALL-CAPS CHARACTER branch below).
  const trait = ws.traits.get(token);
  if (trait) {
    let body = `**TRAIT** \`${token}(${trait.params.join(", ")})\``;
    if (trait.beats.length > 0) body += `\n\nships: ${trait.beats.join(", ")}`;
    return markdown(body, range);
  }

  // CHARACTER reference: ALL-CAPS token that matches a declared name.
  if (allChars(token, (c) => isAsciiUppercase(c) || c === "_")) {
    const info = ws.characters.get(token);
    if (info) {
      let body = `**CHARACTER** \`${token}\``;
      if (info.mixins.length > 0) body += `\n\nis ${info.mixins.join(", ")}`;
      if (info.body.length > 0) {
        body += "\n\n```loom\n";
        for (const bodyLine of info.body) {
          body += bodyLine;
          body += "\n";
        }
        body += "```";
      }
      return markdown(body, range);
    }
  }

  return null;
}

function markdown(value: string, range?: Range): Hover {
  return {
    contents: { kind: "markdown", value },
    ...(range ? { range } : {}),
  };
}
